ComboBox blog post (#1260)
adobe/react-spectrum: packages/dev/docs/pages/blog/building-a-combobox.mdx

OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

import comboboxAccessibilityUrl from 'url:../assets/combobox-accessibility.mp4';
import comboboxScrollingUrl from 'url:../assets/combobox-scrolling-safari.mp4';
import comboboxVisualViewportUrl from 'url:../assets/combobox-visual-viewport.mp4';
import comboboxExampleImageUrl from 'url:../assets/combobox-example.png';
import screenreaderSpreadsheetImageUrl from 'url:../assets/screenreader-spreadsheet-combobox.png';

import {BlogPostLayout, Video, Image} from '@react-spectrum/docs';
export default BlogPostLayout;

---
keywords: [combobox, accessibility, mobile, react spectrum, react, spectrum, interactions, touch]
description: In this post we take a deeper look into some of the challenges faced when building an accessible and mobile friendly ComboBox.
date: 2021-01-08
author: '[Daniel Lu](https://github.com/LFDanLu)'
---

# Building a ComboBox

<Image src={comboboxExampleImageUrl} alt="Screenshot of example ComboBox, select a major" style={{maxWidth: 'min(50%, 350px)', display: 'block', margin: '20px auto'}} />

> As of November 30th, [React Spectrum ComboBox](../react-spectrum/ComboBox.html) has been released as an alpha. For those who may be unfamiliar with a combobox, it is an input field that has a popover menu associated with it. The popover menu contains

Reviewer: ```suggestion
The [React Spectrum ComboBox](../react-spectrum/ComboBox.html) was released as an alpha on November 30th. For review of its functionality and purpose, a combobox is an input field that has a popover menu associated with it. The popover menu contains
```

Author: I feel like the tone change in the second sentence is a bit too formal, but I'll change the first one

<Video src={comboboxVisualViewportUrl} loop autoPlay muted style={{maxWidth: 'min(30%, 350px)', display: 'block', margin: '20px auto'}} />

### Page Scrolling

Another issue we encountered had to do with iOS Safari page scrolling behavior. When the onscreen keyboard is visible, iOS Safari makes the page scrollable so that users can still access content that is hidden behind the keyboard. However,
now that our ComboBox tray sizes itself to fit in the visual viewport, users could now scroll the entire tray itself off screen. To stop this from happening, we prevent default on `touchmove` events that happen on the document body
or root element of the document. This preserves the user's ability to scroll through the options in the tray but blocks any attempt to scroll the page itself until the tray is closed. The video below
illustrates the difference in scrolling behavior before and after our fix. If you are building your own overlays and would like to prevent this kind of document scrolling behavior, check out the [usePreventScroll](../react-aria/usePreventScroll.html) hook in the `react-aria/overlays` package.

> <Video src={comboboxScrollingUrl} loop autoPlay muted style={{maxWidth: 'min(30%, 350px)', display: 'block', margin: '20px auto'}} />

Reviewer: Do we want a video or image showing it working after the changes?

Author: The video actually includes a before/after, just may not be super apparent

so that we have control over how the message should be announced to the user. Then whenever updates occur in the ComboBox, we can have the screen reader announce custom messages by updating the contents of said `div` elements.

Special care was taken such that the messages themselves only contained relevant and concise information tailored to the current state of the ComboBox at any given time. For example, the ComboBox only announces the current section name and item count when the user enters a new section in the menu.
When the user then moves to a different option in the same section, only the newly focused item name is announced. Similarly, the total option count is only announced when number of options available in the menu changes. Since many of these messages were added to fill in gaps in VoiceOver's announcement,
we only trigger the LiveAnnouncer on Apple devices to avoid announcement overlap on other screen readers.

> If you are interested in using this LiveAnnouncer yourself, check out `LiveAnnouncer` in `@react-aria/live-announcer`. Otherwise, the [useComboBox](../react-aria/useComboBox.html) hook provides you with all of the custom messaging out of the box. See the video below for a sneak peek!

Reviewer: I felt like we described the issues and the functionality to create the behavior to fix this, but don't smoothly say we call this LiveAnnouncer. Also, would it be useful to identify a non-combobox use for LiveAnnouncer. Why do we have LiveAnnouncer with and without backticks in that sentence?

<Video src={comboboxAccessibilityUrl} alt="Demo of VoiceOver announcement in ComboBox" style={{maxWidth: 'min(90%, 700px)', display: 'block', margin: '20px auto'}} controls />

## Conclusion

> When we first set off to build ComboBox, we had no idea of the complexity that awaited us. If you're building your own combo box component for your design system, I'd recommend checking out the [useComboBox](../react-aria/useComboBox.html) hook which handles all of this complexity and has been tested across a wide variety of devices, browsers, and assistive technology combinations.

Reviewer: Just workshopping a different way to encourage people to use this. `The useComboBox hook we've built in react-aria is a great tool to build your own ComboBox or add to any ComboBox that doesn't have this mobile and accessibility functionality. It handles all of...`

Author: IMO I like the flow of the original because it follows a "We built this -> If you are building this, check out what we build" structure

import comboboxScrollingUrl from 'url:../assets/combobox-scrolling-safari.mp4';
import comboboxVisualViewportUrl from 'url:../assets/combobox-visual-viewport.mp4';
import comboboxExampleImageUrl from 'url:../assets/combobox-example.png';
import screenreaderSpreadsheetImageUrl from 'url:../assets/screenreader-spreadsheet-combobox.png';

import {BlogPostLayout, Video, Image} from '@react-spectrum/docs';
export default BlogPostLayout;

---
keywords: [combobox, accessibility, mobile, react spectrum, react, spectrum, interactions, touch]
description: In this post we take a deeper look into some of the challenges faced when building an accessible and mobile friendly ComboBox.
date: 2021-01-08
author: '[Daniel Lu](https://github.com/LFDanLu)'
---

# Building a ComboBox

<Image src={comboboxExampleImageUrl} alt="Screenshot of example ComboBox, select a major" style={{maxWidth: 'min(50%, 350px)', display: 'block', margin: '20px auto'}} />

The [React Spectrum ComboBox](../react-spectrum/ComboBox.html) is now officially out of alpha as of January 14! For those who may be unfamiliar with a combobox, it is an input field that has a popover menu associated with it. The popover menu contains

> a list of values within that a user may select as the value for the combobox and is often filtered to display possible matches to the user's current input text. Finally, there is a button accompanying the input field used

Reviewer 2: I think I'd maybe add some examples of where people might have seen combo boxes before. People might know the term "autocomplete" more, so examples like google search would be good.

> a list of values within that a user may select as the value for the combobox and is often filtered to display possible matches to the user's current input text. Finally, there is a button accompanying the input field used

Reviewer 2: ```suggestion
a list of options within that a user may select as the value for the combobox and is often filtered to display possible matches to the user's current input text. Finally, there is a button accompanying the input field used
```

> for toggling the open state of the popover menu.

Reviewer 2: ```suggestion
for toggling the open state of the popover listbox.
```

While the above may seem rather straight forward, a combobox contains a significant amount of complexity. In this blog post, we'll be covering the challenges we encountered with regards to ComboBox's mobile experience and accessibility.

## Mobile experience

### Onscreen Keyboard

> Like many of our other components, ComboBox displays its popover menu items in a tray when rendered on a mobile device or a smaller screen. This allows for a better mobile experience

Reviewer 2: ```suggestion
Like many of our other components, ComboBox displays its options in a tray rather than a popover when rendered on a mobile device or a smaller screen. This allows for a better mobile experience
```

since the tray can display more items on the screen and grants users a larger hit area to scroll through. To compensate for the fact that the tray covers the majority of the
screen including the ComboBox, we included an input field within the tray so that end users could still type to filter the available items. So far so good. However,
when we tested the ComboBox's tray input on iOS Safari, the tray's bottom half was covered by the onscreen keyboard and we could not scroll the hidden items into view.

It turns out that iOS Safari does not shrink the browser window to accommodate the onscreen keyboard unlike other mobile browsers. Instead, Safari pushes the window upwards and
partially off screen. At the time, we had been relying on `window.innerHeight` to inform us on how tall the tray should be but now we had to find a different way of tracking the available window
space for the tray.

> Luckily for us, iOS 13 added support for the VisualViewport API. By querying `window.VisualViewport.height` we could get a reliable measurement of how much vertical space was available on screen. Furthermore, we could track

Reviewer 2: ```suggestion
Luckily for us, iOS 13 added support for the [VisualViewport](https://developer.mozilla.org/en-US/docs/Web/API/VisualViewport) API. By querying `window.visualViewport.height` we could get a reliable measurement of how much vertical space was available on screen. Furthermore, we could track
```

or root element of the document. This preserves the user's ability to scroll through the options in the tray but blocks any attempt to scroll the page itself until the tray is closed. The video below
illustrates the difference in scrolling behavior before and after our fix. If you are building your own overlays and would like to prevent this kind of document scrolling behavior, check out the [usePreventScroll](../react-aria/usePreventScroll.html) hook in the `react-aria/overlays` package.

<Video src={comboboxScrollingUrl} loop autoPlay muted style={{maxWidth: 'min(30%, 350px)', display: 'block', margin: '20px auto'}} />

## Accessibility

To ensure that our ComboBox is accessible, we followed the [WAI-ARIA 1.2 combobox example](https://www.w3.org/TR/wai-aria-practices-1.2/#combobox). Unfortunately, differences between our implementation and the spec arose
as we built ComboBox and subsequent testing sessions revealed varied support across screen readers. Here are just a couple of the challenges we faced and the solutions we came up with to address them.

### Portals

> One significant difference between the WAI-ARIA 1.2 ComboBox spec and our ComboBox implementation was the usage of React Portals for the menu. In the [example](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-list.html)

Reviewer 2: ```suggestion
One significant difference between the WAI-ARIA 1.2 ComboBox spec and our ComboBox implementation was the usage of React Portals for the popover. In the [example](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-list.html)
```

> provided by the spec, the combobox menu is a sibling to the combobox, allowing for easy navigation between the the two elements when using a screenreader. The same couldn't be said for our ComboBox menu since

Reviewer 2: technically, ARIA practices is not a spec

```suggestion
the listbox is a sibling to the combobox, allowing for easy navigation between the the two elements when using a screenreader. The same couldn't be said for our listbox since
```

> it is portalled to the end of the document, meaning there could be any number of elements that a screenreader user would have to navigate through just to move from the ComboBox input to the menu itself. For touch screen reader users, it would be particularly problematic

Reviewer 2: ```suggestion
it is portalled to the end of the document, meaning there could be any number of elements that a screenreader user would have to navigate through just to move from the ComboBox input to the listbox itself. For touch screen reader users, it would be particularly problematic
```

> since they may not have access to a physical keyboard and thus can't use the arrow keys as to navigate through menu items.

Reviewer 2: ```suggestion
since they may not have access to a physical keyboard and thus can't use the arrow keys as to navigate through options.
```

> This put us in a bind because removing the portal strategy wasn't really an option. If we didn't portal the popover menu out to the end of the document, we'd run into the risk of the combobox's menu being cut off if the combobox's parent has `overflow: hidden` or `overflow: scroll`

Reviewer 2: ```suggestion
This put us in a bind because removing the portal strategy wasn't really an option. If we didn't portal the popover out to the end of the document, we'd run into the risk of the listbox being clipped if the combobox's parent had `overflow: hidden` or `overflow: scroll`
```

> set. Therefore, we had to figure out a way to make the ComboBox input and the ComboBox menu the only elements accessible to screen readers while the menu is open so that a screen reader user wouldn't accidentally leave the ComboBox context.

Reviewer 2: ```suggestion
set. Therefore, we had to figure out a way to make the input and listbox the only elements accessible to screen readers while the popover is open so that a screen reader user wouldn't accidentally leave the ComboBox context.
```

> The solution that we came up with was to crawl the DOM and apply `aria-hidden` to every element that wasn't the ComboBox input or menu. To crawl the DOM we used a [TreeWalker](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker),

Reviewer 2: ```suggestion
The solution that we came up with was to crawl the DOM and apply `aria-hidden` to every element that wasn't the input or listbox. To crawl the DOM we used a [TreeWalker](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker),
```

> setting up a node filter to determine if a node should be left visible, hidden, or skipped in the case where its parent was already hidden. In addition, we watch for any changes in the DOM while the menu is open

Reviewer 2: ```suggestion
setting up a node filter to determine if a node should be left visible, hidden, or skipped in the case where its parent was already hidden. In addition, we watch for any changes in the DOM while the listbox is open
```

> via a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), hiding those new elements if need be. When the menu closes, every node that we modified is reverted back to its previous state.

Reviewer 2: ```suggestion
via a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), hiding those new elements if need be. When the popover closes, every node that we modified is reverted back to its previous state.
```

> Check out our `ariaHideOutside` hook in `@react-aria/overlays` if you'd like to learn more.

Reviewer 2: ```suggestion
Check out our [ariaHideOutside](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/overlays/src/ariaHideOutside.ts) function in `@react-aria/overlays` if you'd like to learn more.
```

### Mobile implementation divergence

> Another divergence between the WAI-ARIA 1.2 ComboBox spec and our ComboBox implementation came as a result of our mobile implementation. Initially, we considered the outer input and the tray input as both being individual combo boxes that mirror each others state. This became increasingly complicated as we

Reviewer 2: ```suggestion
Another divergence between the WAI-ARIA 1.2 ComboBox example and our ComboBox implementation came as a result of our mobile implementation. Initially, we considered the outer input and the tray input as both being individual combo boxes that mirror each other's state. This became increasingly complicated as we
```

since the screen reader would detect that the input had `role="combobox"` and thus erroneously announce "double tap to close", a touch action usually reserved for focusing an input. After some deliberation, we decided to alter the accessibility implementation for the mobile experience, removing the outer combo box input in favor of a button that
resembles a combo box. Tapping the button opens the tray that contains the actual combo box input, albeit with `role="searchbox"` instead of `role="combobox"` to avoid the aforementioned screen reader announcement.

### Differences in screen readers

Just like browsers, screen readers can be quite varied in behavior and may require deliberate tooling to ensure a consistent degree of functionality. While automated accessibility tools can be useful for ensuring that your component meets baseline accessibility guidelines,
manual testing is irreplaceable for surfacing these behavioral differences, illustrated in part by the image below.

<Image src={screenreaderSpreadsheetImageUrl} alt="Screenshot of ComboBox screen reader vs browser test matrix" expandable />

#### NVDA

> Initially, our ComboBox would automatically focus the first item in the menu whenever opened so the user wouldn't have to issue another keypress to begin menu navigation. However, when we tested the ComboBox using NVDA we discovered that

Reviewer 2: ```suggestion
Initially, our ComboBox would automatically focus the first item in the listbox whenever opened so the user wouldn't have to issue another keypress to begin option navigation. However, when we tested the ComboBox using NVDA we discovered that
```

> character deletions and text cursor movement in the ComboBox input weren't being announced at all. NVDA would only resume announcing if no menu items had virtual focus via `aria-activedescendant`, so we ended up clearing menu item focus on any changes to the input text or

Reviewer 2: ```suggestion
character deletions and text cursor movement in the ComboBox input weren't being announced at all. NVDA would only resume announcing if no options had virtual focus via `aria-activedescendant`, so we ended up clearing option focus on any changes to the input text or
```

left/right arrow key presses.

#### VoiceOver

VoiceOver has limited support for `aria-activedescendant`, failing to announce when the focused ComboBox item changed in a variety of situations. In Safari, changes to the `aria-activedescendant` aren't announced
when the input is empty. In Chrome, VoiceOver only announces changes to `aria-activedescendant` if `aria-selected="true"` is applied to the focused item. If the list of options is organized into sections using `role="group"`,
VoiceOver doesn't announce the focused item at all in any browser.

> Announcing item focus wasn't the only issue we encountered. VoiceOver doesn't announce the number of options currently available in the menu or the current section title, information that an end user would find helpful when navigating the ComboBox menu.

Reviewer 2: ```suggestion
Announcing item focus wasn't the only issue we encountered. VoiceOver doesn't announce the number of options currently available in the menu or the current section title, information that a user would find helpful when navigating the ComboBox options.
```

> When selecting an option from the menu, an announcement of the item's name is made but the fact that it is now "selected" is omitted.

Reviewer 2: ```suggestion
When selecting an option from the listbox, an announcement of the item's name is made but the fact that it is now "selected" is omitted.
```

> As a workaround, we made use of [ARIA live-regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions). When the ComboBox's menu is opened for the first time, we create two visually hidden `aria-live` regions in the DOM with two `div` elements, one with `aria-live="polite"` and the other `aria-live="assertive"`

Reviewer 2: ```suggestion
As a workaround, we made use of [ARIA live-regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions). When the ComboBox's listbox is opened for the first time, we create two visually hidden `aria-live` regions in the DOM with two `div` elements, one with `aria-live="polite"` and the other `aria-live="assertive"`
```

> so that we have control over how the message should be announced to the user. Then whenever updates occur in the ComboBox, we can have the screen reader announce custom messages by updating the contents of said `div` elements. Thankfully, the bulk of this functionality was setup already in

Reviewer 2: ```suggestion
so that we have control over how the message should be announced to the user. Then whenever updates occur in the ComboBox, we can have the screen reader announce custom messages by updating the contents of these elements. Thankfully, the bulk of this functionality was setup already in
```

React Aria's `LiveAnnouncer`, allowing us to reuse it for our ComboBox after some updates.

> Special care was taken such that the messages themselves only contained relevant and concise information tailored to the current state of the ComboBox at any given time. For example, the ComboBox only announces the current section name and item count when the user enters a new section in the menu.

Reviewer 2: ```suggestion
Special care was taken such that the messages themselves only contained relevant and concise information tailored to the current state of the ComboBox at any given time. For example, the ComboBox only announces the current section name and item count when the user enters a new section in the listbox.
```

> When the user then moves to a different option in the same section, only the newly focused item name is announced. Similarly, the total option count is only announced when number of options available in the menu changes. Since many of these messages were added to fill in gaps in VoiceOver's announcement,

Reviewer 2: ```suggestion
When the user then moves to a different option in the same section, only the newly focused item name is announced. Similarly, the total option count is only announced when number of options available in the listbox changes. Since many of these messages were added to fill in gaps in VoiceOver's announcement,
```

we only trigger the `LiveAnnouncer` on Apple devices to avoid announcement overlap on other screen readers.

> If you are interested in using this `LiveAnnouncer` yourself, check out `LiveAnnouncer` in `@react-aria/live-announcer`. Otherwise, the [useComboBox](../react-aria/useComboBox.html) hook provides you with all of the custom messaging out of the box. See the video below for a sneak peek!

Reviewer 2: ```suggestion
If you are interested in using this `LiveAnnouncer` yourself, check out [LiveAnnouncer](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/live-announcer/src/LiveAnnouncer.tsx) in `@react-aria/live-announcer`. Otherwise, the [useComboBox](../react-aria/useComboBox.html) hook provides you with all of the custom messaging out of the box. See the video below for a sneak peek!
```

OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

import comboboxAccessibilityUrl from 'url:../assets/combobox-accessibility.mp4';
import comboboxScrollingUrl from 'url:../assets/combobox-scrolling-safari.mp4';
import comboboxVisualViewportUrl from 'url:../assets/combobox-visual-viewport.mp4';
import comboboxExampleImageUrl from 'url:../assets/combobox-example.png';
import screenreaderSpreadsheetImageUrl from 'url:../assets/screenreader-spreadsheet-combobox.png';

import {BlogPostLayout, Video, Image} from '@react-spectrum/docs';
export default BlogPostLayout;

---
keywords: [combobox, accessibility, mobile, react spectrum, react, spectrum, interactions, touch]
description: In this post we take a deeper look into some of the challenges faced when building an accessible and mobile friendly ComboBox.
date: 2020-11-30
author: '[Daniel Lu](https://github.com/LFDanLu)'
---

# Building a ComboBox

## Intro

<Image src={comboboxExampleImageUrl} alt="Screenshot of example ComboBox, select a major" style={{maxWidth: 'min(50%, 350px)', display: 'block', margin: '20px auto'}} />

> As of November 30th, React Spectrum ComboBox has been released as an alpha. For those who may be unfamiliar with a combobox, it is an input field that has a popover menu associated with it. The popover menu contains

Reviewer 3: Add link to ComboBox docs

screen including the ComboBox, we included an input field within the tray so that end users could still type to filter the available items. So far so good. However,
when we tested the ComboBox's tray input on iOS Safari, the tray's bottom half was covered by the onscreen keyboard and we could not scroll the hidden items into view.

It turns out that iOS Safari does not shrink the browser window to accommodate the onscreen keyboard unlike other mobile browsers. Instead, Safari pushes the window upwards and
partially off screen. At the time, we had been relying on `window.innerHeight` to inform us on how tall the tray should be but now we had to find a different way of tracking the available window
space for the tray.

Luckily for us, iOS 13 added support for the VisualViewport API. By querying `window.VisualViewport.height` we could get a reliable measurement of how much vertical space was available on screen. Furthermore, we could track
when the onscreen keyboard was opened or dismissed by listening to the VisualViewport's `resize` event. Leveraging these two allowed us to create a tray that properly adjusts to the presence of iOS onscreen keyboard. Check out the video below

> to see how the ComboBox tray worked before and after we switched to the VisualViewport API. If you'd like to track the visual viewport size in your own app, you can use the `useViewportSize` hook available in the `@react-aria/utils` package.

Reviewer 2: IMO we could just link to the source code.

```suggestion
to see how the ComboBox tray worked before and after we switched to the VisualViewport API. If you'd like to track the visual viewport size in your own app, you can use the [useViewportSize](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/utils/src/useViewportSize.ts) hook available in the `@react-aria/utils` package.
```

set. Therefore, we had to figure out a way to make the ComboBox input and the ComboBox menu the only elements accessible to screen readers while the menu is open so that a screen reader user wouldn't accidentally leave the ComboBox context.

The solution that we came up with was to crawl the DOM and apply `aria-hidden` to every element that wasn't the ComboBox input or menu. To crawl the DOM we used a [TreeWalker](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker),
setting up a node filter to determine if a node should be left visible, hidden, or skipped in the case where its parent was already hidden. In addition, we watch for any changes in the DOM while the menu is open
via a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), hiding those new elements if need be. When the menu closes, every node that we modified is reverted back to its previous state.
Check out our `ariaHideOutside` hook in `@react-aria/overlays` if you'd like to learn more.

### Mobile implementation divergence

> Another divergence between the WAI-ARIA 1.2 ComboBox spec and our ComboBox implementation came as a result of our mobile implementation. Initially, we considered the outer input and the tray input as both being individual combo boxes that mirror each others state. This became increasingly complicated as we tried to

Reviewer 3: This sentence is too long
"This became increasingly complicated as we tried to ..." Can we break it up a bit?

This became increasingly complicated as we tried to manage things like focus and internal state between the two inputs. The result included adding several tray input specific affordances added to our `useComboBox` hook as well as reducing the aria hook's generality by introducing this React Spectrum specific behavior.

manage things like focus and internal state between the two inputs, resulting in several tray input specific affordances added to our `useComboBox` hook, reducing the aria hook's generality by introducing this React Spectrum specific behavior. Interacting with the tray input with a screen reader was also problematic

> since the screen reader would detect that the input had `role="combobox"` and thus erroneously state "double tap to close", a touch action usually reserved for focusing an input. After some deliberation, we decided to alter the accessibility implementation for the mobile experience, removing the outer combo box input in favor for a button that

Reviewer 3: ```suggestion
since the screen reader would detect that the input had `role="combobox"` and thus erroneously announce "double tap to close", a touch action usually reserved for focusing an input. After some deliberation, we decided to alter the accessibility implementation for the mobile experience, removing the outer combo box input in favor of a button that
```

> looks like a combo box. Tapping the button opens the tray that contains the actual combo box input, albeit with `role="searchbox"` instead of `role="combobox"` to avoid the aforementioned screen reader announcement.

Reviewer 3: ```suggestion
resembled a combo box. Tapping the button opens the tray that contains the actual combo box input, albeit with `role="searchbox"` instead of `role="combobox"` to avoid the aforementioned screen reader announcement.
```

### Differences in screen readers

Just like browsers, screen readers can be quite varied in behavior and may require deliberate tooling to ensure a consistent degree of functionality. While automated accessibility tools can be useful for ensuring that your component meets baseline accessibility guidelines,
manual testing is irreplaceable for surfacing these behavioral differences, illustrated in part by the image below.

> <Image src={screenreaderSpreadsheetImageUrl} alt="Screenshot of ComboBox screen reader vs browser test matrix" style={{maxWidth: 'min(100%, 780px)', display: 'block', margin: '20px auto'}} />

Reviewer 4: this is hard to read because i think it's being resized, unless I zoom the screen i really can't read it at all :(
Is there any chance of including it as a real table so is unaffected by scaling?

Author: Hm, I can try that I guess, maybe I could make it so when you click on the image it expands?

Reviewer 4: maybe, that's usually a whole component that handles it, usually like a lightbox i think

#### NVDA

Initially, our ComboBox would automatically focus the first item in the menu whenever opened so the user wouldn't have to issue another keypress to begin menu navigation. However, when we tested the ComboBox using NVDA we discovered that
character deletions and text cursor movement in the ComboBox input weren't being announced at all. NVDA would only resume announcing if no menu items had virtual focus via `aria-activedescendant`, so we ended up clearing menu item focus on any changes to the input text or
left/right arrow key presses.

#### VoiceOver

> VoiceOver as a whole doesn't seem to work very well with `aria-activedescendant`, failing to announce when the focused ComboBox item changed in a variety of situations. In Safari, changes to the `aria-activedescendant` aren't announced

Reviewer 3: Maybe VoiceOver has limited support for `aria-activedescendant`...

so that we have control over how the message should be announced to the user. Then whenever updates occur in the ComboBox, we can have the screen reader announce custom messages by updating the contents of said `div` elements.

Special care was taken such that the messages themselves only contained relevant and concise information tailored to the current state of the ComboBox at any given time. For example, the ComboBox only announces the current section name and item count when the user enters a new section in the menu.
When the user then moves to a different option in the same section, only the newly focused item name is announced. Similarly, the total option count is only announced when number of options available in the menu changes. Since many of these messages were added to fill in gaps in VoiceOver's announcement,
we only trigger the LiveAnnouncer on Apple devices to avoid announcement overlap on other screen readers.

If you are interested in using this LiveAnnouncer yourself, check out `LiveAnnouncer` in `@react-aria/live-announcer`. Otherwise, the `useComboBox` hook provides you with all of the custom messaging out of the box. See the video below for a sneak peek!

> <Video src={comboboxAccessibilityUrl} alt="Demo of VoiceOver announcement in ComboBox" style={{maxWidth: 'min(90%, 700px)', display: 'block', margin: '20px auto'}} />

Reviewer 4: this video doesn't play for me?

Author: yeaaaaaah, I prevented autoplay since I hate it when websites have an autoplay video with audio. Lemme play around with some more settings to make this more friendly, but you should be able to right click it and display the video controls and play the video that way

Reviewer 4: ahhhh, i would never have thought of that. but yes, it does work. Agreed, hate autoplay, but there has to be some way to show the controls by default?

## Conclusion

> When we first set off to build ComboBox, we had no idea of the complexity that awaited us. If you're building your own combo box component for your design system, I'd recommend checking out the `useComboBox` hook which handles all of this complexity and has been tested across a wide variety of devices, browsers, and assistive technology combinations.

Reviewer 3: Add docs link(s)

<Video src={comboboxScrollingUrl} loop autoPlay muted style={{maxWidth: 'min(30%, 350px)', display: 'block', margin: '20px auto'}} />

## Accessibility

To ensure that our ComboBox is accessible, we followed the [WAI-ARIA 1.2 combobox example](https://www.w3.org/TR/wai-aria-practices-1.2/#combobox). Unfortunately, differences between our implementation and the spec arose
as we built ComboBox and subsequent testing sessions revealed varied support across screen readers. Here are just a couple of the challenges we faced and the solutions we came up with to address them.

### Portals

One significant difference between the WAI-ARIA 1.2 ComboBox spec and our ComboBox implementation was the usage of React Portals for the listbox. In the [example](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-list.html)
the listbox is a sibling to the combobox, allowing for easy navigation between the the two elements when using a screenreader. The same couldn't be said for our listbox since

> it is portalled to the end of the document, meaning there could be any number of elements that a screenreader user would have to navigate through just to move from the ComboBox input to the listbox itself. For touch screen reader users, it would be particularly problematic

Reviewer 3: ```suggestion
it is portalled to the end of the document, meaning there could be any number of elements that a screen reader user would have to navigate through just to move from the ComboBox input to the listbox itself. For touch screen reader users, it would be particularly problematic
```

setting up a node filter to determine if a node should be left visible, hidden, or skipped in the case where its parent was already hidden. In addition, we watch for any changes in the DOM while the listbox is open
via a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), hiding those new elements if need be. When the popover closes, every node that we modified is reverted back to its previous state.
Check out our [ariaHideOutside](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/overlays/src/ariaHideOutside.ts) function in `@react-aria/overlays` if you'd like to learn more.

### Mobile implementation divergence

Another divergence between the WAI-ARIA 1.2 ComboBox example and our ComboBox implementation came as a result of our mobile implementation. Initially, we considered the outer input and the tray input as both being individual combo boxes that mirror each other's state. This became increasingly complicated as we

> tried to manage things like focus and internal state between the two inputs. The result included adding several tray input specific affordances added to our `useComboBox` hook as well as reducing the aria hook's generality by introducing this React Spectrum specific behavior. Interacting with the tray input with a screen reader was also problematic

Reviewer 3: ```suggestion
tried to manage things like focus and internal state between the two inputs. The result included adding several tray input specific affordances to our `useComboBox` hook as well as reducing the aria hook's generality by introducing this React Spectrum specific behavior. Interacting with the tray input using a screen reader was also problematic
```

### Differences in screen readers

Just like browsers, screen readers can be quite varied in behavior and may require deliberate tooling to ensure a consistent degree of functionality. While automated accessibility tools can be useful for ensuring that your component meets baseline accessibility guidelines,
manual testing is irreplaceable for surfacing these behavioral differences, illustrated in part by the image below.

```tsx import
import screenreaderSpreadsheetImageUrl from 'url:../assets/screenreader-spreadsheet-combobox.png';
import {Image, ImageContext, ImageModal} from '@react-spectrum/docs';

ReactDOM.render((
  <ImageContext.Provider value="">
    <ImageModal>
      <Image src={screenreaderSpreadsheetImageUrl} alt="Screenshot of ComboBox screen reader vs browser test matrix" style={{maxWidth: 'min(100%, 780px)', display: 'block', margin: '20px auto', cursor: 'zoom-in'}} />
      <Image src={screenreaderSpreadsheetImageUrl} alt="Screenshot of ComboBox screen reader vs browser test matrix" style={{maxWidth: 'initial'}} />
    </ImageModal>
  </ImageContext.Provider>

> ), document.getElementById('spreadsheet-image'));

Author: shoot looks like this doesn't work in the built version, gonna try doing something with portals perhaps

### Differences in screen readers

Just like browsers, screen readers can be quite varied in behavior and may require deliberate tooling to ensure a consistent degree of functionality. While automated accessibility tools can be useful for ensuring that your component meets baseline accessibility guidelines,
manual testing is irreplaceable for surfacing these behavioral differences, illustrated in part by the image below.

```tsx import
import screenreaderSpreadsheetImageUrl from 'url:../assets/screenreader-spreadsheet-combobox.png';
import {Image, ImageContext, ImageModal} from '@react-spectrum/docs';

function Blah() {
  let publicUrl = React.useContext(ImageContext);
  return (
    <ImageContext.Provider value={publicUrl || ''}>
      <ImageModal>
        <Image src={screenreaderSpreadsheetImageUrl} alt="Screenshot of ComboBox screen reader vs browser test matrix" style={{maxWidth: 'min(100%, 780px)', display: 'block', margin: '20px auto', cursor: 'zoom-in'}} />
        <Image src={screenreaderSpreadsheetImageUrl} alt="Screenshot of ComboBox screen reader vs browser test matrix" style={{maxWidth: 'initial'}} />
      </ImageModal>

>     </ImageContext.Provider>

Author: I'm doing something horribly wrong here, ugh

<!-- Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

import {BlogPostLayout} from '@react-spectrum/docs';
export default BlogPostLayout;

---
keywords: [combobox, accessibility, mobile, react spectrum, react, spectrum, interactions, touch]
description: In this post we take a deeper look into some of the challenges faced when building an accessible and mobile friendly ComboBox.

> date: 2020-08-25

Author: Will update date once it goes out

author: '[Daniel Lu](https://github.com/LFDanLu)'
---

# Building a ComboBox

## Intro

> As of INSERT DATE HERE, React Spectrum ComboBox has been released as an alpha. For those who may be unfamiliar with a combobox, it is an input field that has a popover menu associated with it. The popover menu contains

Author: I figured this would go up after combobox releases as an alpha, hence the placeholder.

when we tested the ComboBox's tray input on iOS Safari, the tray's bottom half was covered by the onscreen keyboard and we could not scroll the hidden items into view.

It turns out that iOS Safari does not shrink the browser window to accommodate the onscreen keyboard unlike other mobile browsers. Instead, Safari pushes the window upwards and
partially off screen. At the time, we had been relying on `window.innerHeight` to inform us on how tall the tray should be but now we had to find a different way of tracking the available window
space for the tray.

Luckily for us, iOS 13 added support for the VisualViewport API. By querying `window.VisualViewport.height` we could get a reliable measurement of how much vertical space was available on screen. Furthermore, we could track
when the onscreen keyboard was opened or dismissed by listening to the VisualViewport's `resize` event. Leveraging these two allowed us to create a tray that properly adjusts to the presence of iOS onscreen keyboard.
If you'd like to track the visual viewport size in your own app, you can use the `useViewportSize` hook available in the `@react-aria/utils` package.

> 

Reviewer 2: I suppose we should also cover the challenges in getting mobile Safari to not scroll the window when focusing the input? The comment in usePreventScroll covers it pretty well.

Author: I can mention preventing page scrolling so that the tray itself can't be scrolled up/down when the keyboard is expanded, but I can't repro the following comment in the combobox after commenting out the fix:

```
3. When tapping on an input, the page always scrolls so that the input is centered in the visual viewport.
    This may cause even fixed position elements to scroll off the screen.
``` 

I think the above was perhaps specific to if the input is near the bottom of the screen prior to the onscreen keyboard being opened (can only repro in the dialog with long content story)

## Accessibility

To ensure that our ComboBox is accessible, we followed the [WAI-ARIA 1.2 combobox spec](https://www.w3.org/TR/wai-aria-practices-1.2/#combobox). Unfortunately, differences between our implementation and the spec arose
as we built ComboBox and subsequent testing sessions revealed varied support across screen readers. Here are just a couple of the challenges we faced and the solutions we came up with to address them.

### Portals

One significant difference between the WAI-ARIA 1.2 ComboBox spec and our ComboBox implementation was the usage of React Portals for the menu. In the [example](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-list.html)
provided by the spec, the combobox menu is a sibling to the combobox, allowing for easy focus navigation between the the two elements when using a screenreader. The same couldn't be said for our ComboBox menu since

> it is portalled to the end of the document, meaning there could be any number of elements that a screenreader user would have to navigate through just to move from the ComboBox input to the menu itself. Furthermore, we had to figure out

Author: Could you expand on the touch screen reader bit? If we didn't portal the menu, screen reader users would use swipe gestures to cycle through the menu even if it was visually cut off by a overflow hidden/scroll parent right? Is the argument there that it would be more troublesome since that is a bit more arduous that just being able to tap on an option to focus it?

a way to make the ComboBox input and the ComboBox menu the only elements accessible to screen readers while the menu is open so that a screen reader user wouldn't accidentally leave the ComboBox context.

The solution that we came up with was to crawl the DOM and apply `aria-hidden` to every element that wasn't the ComboBox input or menu. To crawl the DOM we used a [TreeWalker](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker),
setting up a node filter to determine if a node should be left visible, hidden, or skipped in the case where its parent was already hidden. In addition, we watch for any changes in the DOM while the menu is open
via a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), hiding those new elements if need be. When the menu closes, every node that we modified is reverted back to its previous state.
Check out our `ariaHideOutside` hook in `@react-aria/overlays` if you'd like to learn more.

> 

Reviewer 2: Maybe add something about the difference in accessibility implementation when using this pattern? How the outer combobox is really a button that opens a dialog (tray) containing the real combobox? This after testing using two comboboxes and finding it challenging to manage focus

Author: Sure I'll add this as a separate section

> ### Differences in screen readers

Author: Decided to cover only NVDA and VoiceOver stuff here out of consideration for length. Lemme know if its too short and any other topics or additions to add here.

Some other things I have noted down that I could briefly discuss:
- Talkback double tap on inputs don't fire clicks (previously was issue when outer combobox was input)
- iPad voiceover double tap fires two touchends (doesn't happen in iPhone)
- using aria-setsize and counting non-section items to get the correct count announced
- changing input in tray from role combobox to searchbox

VoiceOver doesn't announce changes to the focused item for all browsers.

Announcing item focus wasn't the only issue encountered. VoiceOver doesn't announce the number of options currently available in the menu or the current section title, information that an end user would find helpful when navigating the ComboBox menu.
When selecting an option from the menu, an announcement of the item's name is made but the fact that it is now "selected" is omitted.

As a workaround, we made use of [ARIA live-regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions). When the ComboBox's menu is opened for the first time, we create two visually hidden `aria-live` regions in the DOM with two `div`s, one with `aria-live="polite"` and the other `aria-live="assertive"`
so that we have control over how the message should be announced to the user. Then whenever updates occur in the ComboBox, we can have the screen reader announce custom messages by updating the contents of said `div`s. This allows us to fill in the announcement gaps mentioned previously as well as standardize the

> announcement experience over each of the browsers. If you are interested in using this LiveAnnouncer yourself, check out `LiveAnnouncer` in `@react-aria/live`.

Author: Sounds good, gonna omit the localization bit since we don't have translations just yet

> 

Author: If there are any tickets filed that I should mention lemme know

> ## Conclusion

Reviewer 3: is that something we want to advertise ?

> ## Conclusion

Author: Yeah, that particular workaround felt particularly iffy so I omitted it.

If there is one takeaway, it is that cross browser and device testing is extremely important when building a component. A solution that may work in one browser and device combo may result in a broken experience in another. Browsers, devices, and assistive technologies each have their own behavioral quirks that often are only made apparent

> after thorough testing.

Author: My own personal takeaway (albeit feels a bit obvious). Kinda wanted to say something along the line of 'combobox is hard' but ehh

Reviewer 2: Yeah I think saying combobox is harder than it looks is reasonable as well. 😉 

You should also add something like "The useComboBox hook handles all of this complexity and has been tested across a wide variety of devices, browsers, and assistive technology combinations. If you're building your own combo box component for your design system, check it out!". Feel free to re-word.
